src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaCourseService.java

```diff
@@ -259,6 +259,27 @@ public void deleteMultiple(List<Long> courseIds) {
 		}
 	}
 
+	@Override
+	public void massAddTagsToCourses(List<Long> tagsToAdd, List<Long> tagsToRemove, List<Long> courseIds) {
+		for (Long courseId : courseIds) {
+			Course course = this.getOneById(courseId);
+
+			for (Long tagId: tagsToAdd) {
+				if (course.getTagIds().indexOf(tagId) == -1) {
+					Tag tag = tagService.getOneById(tagId);
+					this.addTag(course, tag);
```

Comment:
Avoiding duplicate tags seems like behavior that belongs in addTag() and removeTag(), not outside of it.